File: 0207-course-schedule/0207-course-schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function (numCourses, prerequisites) {
    if (numCourses === 1 && prerequisites.length > 0) return false;

    const adjList = new Map();
    const inDegree = new Map();
    for (let i = 0; i < numCourses; i++) {
        adjList.set(i, []);
        inDegree.set(i, 0);
    }

    for (const [to, from] of prerequisites) {
        const list = adjList.get(from);
        list.push(to);
        adjList.set(from, list);
        inDegree.set(to, inDegree.get(to) + 1);
    }

    console.log(adjList, inDegree);

    const q = [];
    for (let i = 0; i < inDegree.size; i++) {
        if (inDegree.get(i) === 0) {
            q.push(i);
        }
    }

    if (q.length === 0) {
        return false;
    }

    let completed = 0;

    while (q.length > 0) {
        const node = q.shift();

        completed++;

        for (let neighbor of adjList.get(node)) {
            inDegree.set(neighbor, inDegree.get(neighbor) - 1);
            if (inDegree.get(neighbor) === 0) {
                q.push(neighbor);
            }
        }
    }

    return completed === numCourses;
};

File: 0207-course-schedule/0207-course-schedule.py
from collections import deque

class Solution:
    def canFinish(self, numCourses: int, prerequisites: list[list[int]]) -> bool:
        adj_list = {}
        in_degree = {}

        for i in range(numCourses):
            adj_list[i] = []
            in_degree[i] = 0

        for to, frm in prerequisites:
            lst = adj_list.get(frm, [])
            lst.append(to)
            in_degree[to] = in_degree.get(to, 0) + 1

        q = deque()
        for key, val in in_degree.items():
            if val == 0:
                q.append(key)

        count =0
        while q:
            node = q.popleft()
            count+=1

            for neighbor in adj_list.get(node, []):
                in_degree[neighbor] = in_degree.get(neighbor) - 1
                if in_degree[neighbor] == 0:
                    q.append(neighbor)

        return count == numCourses